wpep0008.md

```diff
@@ -0,0 +1,122 @@
+| Key | Value |
+| ---: | :--- |
+| **Title** | Wazo integration tests framework |
+| **Number** | 0008 |
+| **Date** | 2023-08-28 |
+| **Authors** | Jesse Sopel <jesse.sopel@wazo.io>, Sebastien Duthil <sduthil@wazo.io> |
+| **Status** | draft |
+
+## Abstract
+
+Make explicit the desire to migrate towards a more pytest-oriented paradigm in
+integration tests.
+
+## Motivation
+
+The integration test framework was implemented in Wazo before the usage of
+pytest. Pytest offers a very Pythonic way to organize tests and fixtures, and we
+want to use this system more and more, since it offers more flexibility,
+readability, performance and reusability.
+
+## State of the art in Wazo Platform
+
+Integration tests are organised around `AssetLaunchingTestCase`, inheriting from
+the `unittest` testing framework. This organization forces tests to be organized
+around a test class, which may not be as flexible as we want.
+
+Here is an example:
+
+```python
+# helpers/base.py
+class BaseAssetLaunchingTestCase(AssetLaunchingTestCase):
+
+    assets_root = os.path.join(os.path.dirname(__file__), '../..', 'assets')
+    asset = 'base'
+    service = 'agid'
+
+    @classmethod
+    def make_agid(cls):
+        port = cls.service_port(4573, 'agid')
+        return AgidClient('127.0.0.1', port)
+
+class IntegrationTest(unittest.TestCase):
+    asset_cls = BaseAssetLaunchingTestCase
+    agid: AgidClient
+
+    @classmethod
+    def setUpClass(cls):
+        cls.agid = cls.asset_cls.make_agid()
+
+# conftest.py
+from .helpers import base as asset
+
+def pytest_collection_modifyitems(session, config, items):
+    item.parent.own_markers[0].args[0])
```

Comment:
```suggestion
def pytest_collection_modifyitems(session, config, items):
    item.parent.own_markers[0].args[0]
```
What does this do?

Reply:
Sorry, I incorrectly copied that line. It was:
```python
items.sort(key=lambda item: item.parent.own_markers[0].args[0])
```
It seems to be a hack to sort the tests by asset name to ensure they are executed in order. However, when using actual pytest fixtures and not classes this hack is not needed as pytest already groups tests by fixtures and cleans up afterwards.

Reply:
Code updated.